# patch for PALASM1-ARCHIVE
# 
# This file is part of PALASM1-ARCHIVE
# 
# PALASM1-ARCHIVE is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# PALASM1-ARCHIVE is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with PALASM1-ARCHIVE.  If not, see <https://www.gnu.org/licenses/>.
#
# Author: Andreas Heitmann
# Date: 2025-01-10
#
diff --strip-trailing-cr --unified orig/palasm20.for new/palasm20.for
--- orig/palasm20.for	2025-01-10 16:07:13.339260300 +0100
+++ new/palasm20.for	2025-01-10 16:07:13.459248700 +0100
@@ -131,6 +131,8 @@
 C
 C***********************************************************************
 C
+C                       PATCHES FOR GFORTRAN COMPILER
+C                       (C) 2024 ANDREAS HEITMANN
 C
 C***********************************************************************
 C
@@ -142,12 +144,15 @@
      1        LFIX,LFIRST,LFUSES(32,64),LPHASE(20),LBUF(20),
      2        LPROD(80),LSAME,LACT,LOPERR,LINP,LPRD,LERR,LSA11,LSA01,
      3        LPHANT(32,64)
-      INTEGER   IPAGE(80,200),IPAL(4),IOT,REST(73),IOP,INOAI,INOO
+      CHARACTER   IPAGE(80,200),IPAL(4),IOT,REST(73),IOP,INOAI,INOO
      1,         E,O,T,P,B,H,S,L,N,C,Q,U,F,R,A,Y,X,IBUF(8,20)
      2,         BB,CC,DD,EE,FF,II,NN,OO,PP,RR,SS,TT,UU,hhh,lll,ppp,nnn
      3,         ISYM(8,20)
      4,         PATNUM(80),TITLE(80),COMP(80)
+      CHARACTER JJ
       INTEGER BEL
+      LOGICAL IFIX
+      CHARACTER(LEN=1024) :: FILENAME
       COMMON  LBLANK,LLEFT,LAND,LOR,LSLASH,LEQUAL,LRIGHT,LXOR,LXNOR
       COMMON /PGE/ IPAGE
       COMMON /LFUSES/LFUSES,LPHANT
@@ -179,6 +184,13 @@
 C
 C
       RPD=1
+C     PATCH: GET INPUT FILE NAME FROM COMMAND LINE
+      CALL GETARG(1, FILENAME)
+      OPEN(UNIT=RPD, FILE=FILENAME, STATUS='OLD', IOSTAT=ios)
+      IF (IOS /= 0) THEN
+      	 PRINT *, 'Error opening file: ', filename
+	 STOP
+      END IF
       ROC=5
       POF=LUN
       PDF=LUN
@@ -466,10 +478,10 @@
 C**************************************************
 C
 C
-      INTEGER FUNCTION ICONV(K)
+      CHARACTER FUNCTION ICONV(K)
       IMPLICIT INTEGER (A-Z)
-C      INTEGER A,B,C,D,E,F,G,H
-C     1,I,J,X,L,M,N,O,P
+      CHARACTER A,B,C,D,E,F,G,H
+     1,I,J,X,L,M,N,O,P
       DATA A/'0'/,B/'1'/,C/'2'/,D/'3'/,E/'4'/,F/'5'/,G/'6'/,H/'7'/
       DATA I/'8'/,J/'9'/,X/'A'/,L/'B'/,M/'C'/,N/'D'/,O/'E'/,P/'F'/
       IF(K.EQ.0) ICONV=A
@@ -501,8 +513,9 @@
       IMPLICIT INTEGER (A-Z)
       LOGICAL LBLANK,LLEFT,LAND,LOR,LSLASH,LEQUAL,LRIGHT,LXOR,LXNOR,
      1        LFIX,LFUSES(32,64),LPHANT(32,64)
-      INTEGER H,L,C,R,X,A,IPAGE(80,200),IOT,INOAI,INOO
-     1         ,I0,I2,I4,I6,I8,TSTVEC(20,50)
+      CHARACTER IPAGE(80,200)
+      CHARACTER TSTVEC(20,50)
+      CHARACTER I8,I6,I4,I2,I0,A,X,R,C,L,H,INOAI,INOO,IOT
       COMMON  LBLANK,LLEFT,LAND,LOR,LSLASH,LEQUAL,LRIGHT,LXOR,LXNOR
       COMMON /PGE/ IPAGE
       COMMON  /TSTVEC/ NTEST, TSTVEC
@@ -553,7 +566,7 @@
       IMPLICIT INTEGER (A-Z)
       LOGICAL LBLANK,LLEFT,LAND,LOR,LSLASH,LEQUAL,LRIGHT,LXOR,LXNOR,
      1        LFIX,LPHASE(20)
-      INTEGER IBLANK,IPAGE(80,200)
+      CHARACTER IBLANK,IPAGE(80,200)
      1,         ISYM(8,20)
       COMMON  LBLANK,LLEFT,LAND,LOR,LSLASH,LEQUAL,LRIGHT,LXOR,LXNOR
       COMMON /PGE/ IPAGE
@@ -586,13 +599,15 @@
       IMPLICIT INTEGER (A-Z)
       LOGICAL LBLANK,LLEFT,LAND,LOR,LSLASH,LEQUAL,LRIGHT,LXOR,LXNOR,
      1        LFIX,LX1
-      INTEGER IBLANK,ILEFT,IAND,IOR,COMENT,IPAGE(80,200)
+      CHARACTER IBLANK,ILEFT,IAND,IOR,COMENT,IPAGE(80,200)
      1         ,ISLASH,IEQUAL,IRIGHT,ICOLON
+      CHARACTER TAB
+      CHARACTER, PARAMETER :: TABP = CHAR(9)
       COMMON  LBLANK,LLEFT,LAND,LOR,LSLASH,LEQUAL,LRIGHT,LXOR,LXNOR
       COMMON /PGE/ IPAGE
       COMMON /LUNIT/ PMS,POF,PDF
       DATA IBLANK/' '/,ILEFT/'('/,IAND/'*'/,IOR/'+'/,COMENT/';'/,
-     1     ISLASH/'/'/,IEQUAL/'='/,IRIGHT/')'/,ICOLON/':'/,TAB/009/
+     1     ISLASH/'/'/,IEQUAL/'='/,IRIGHT/')'/,ICOLON/':'/,TAB/TABP/
       LBLANK=.FALSE.
       LXOR=.FALSE.
       LXNOR=.FALSE.
@@ -641,7 +656,7 @@
 C      AND THE PIN NAME IN THE PIN LIST OR FUNCTION TABLE PIN LIST
       IMPLICIT INTEGER (A-Z)
       LOGICAL LMATCH
-      INTEGER C,A,R,Y,IBUF(8,20),ISYM(8,20)
+      CHARACTER C,A,R,Y,IBUF(8,20),ISYM(8,20)
       DATA C/'C'/,A/'A'/,R/'R'/,Y/'Y'/
       IMATCH=0
       DO 20 J=1,20
@@ -688,10 +703,13 @@
 C      PAL16X4 AND PAL16A4
       IMPLICIT INTEGER (A-Z)
       LOGICAL LBUF(20),LFUSES(32,64),LFIRST,LMATCH,LFIX
-      INTEGER A,B,ISLASH,IOR,IBLANK,IRIGHT,IPAGE(80,200)
+      LOGICAL LPROD(80)
+      CHARACTER IOP
+      CHARACTER A,B,ISLASH,IOR,IBLANK,IRIGHT,IPAGE(80,200)
      1         ,IAND,N,Q,N0,N1,N2,N3,CI
      2         ,ICOLON
      3         ,TABLE(5,14),IBUF(8,20),FIXBUF(8)
+      CHARACTER TITLE
       COMMON /PGE/ IPAGE
       DATA A/'A'/,B/'B'/,ISLASH/'/'/,IOR/'+'/,IBLANK/' '/,IRIGHT/')'/,
      1        IAND/'*'/,N/'N'/,Q/'Q'/,N0/'0'/,N1/'1'/,N2/'2'/,N3/'3'/,
@@ -754,7 +772,7 @@
       SUBROUTINE ECHO(IPAL,INOAI,IOT,INOO,REST,PATNUM,TITLE,COMP)
 C     THIS SUBROUTINE PRINTS THE PAL DESIGN SPECIFICATION INPUT FILE
       IMPLICIT INTEGER (A-Z)
-      INTEGER IBLANK,IPAGE(80,200),IPAL(4),IOT,INOAI,INOO
+      CHARACTER IBLANK,IPAGE(80,200),IPAL(4),IOT,INOAI,INOO
      1,         PATNUM(80),TITLE(80),COMP(80),REST(73)
       COMMON /PGE/ IPAGE
       COMMON /LUNIT/ PMS,POF,PDF
@@ -813,7 +831,7 @@
       SUBROUTINE PINOUT(IPAL,INOAI,IOT,INOO,TITLE)
 C     THIS SUBROUTINE PRINTS THE PINOUT OF THE PAL
       IMPLICIT INTEGER (A-Z)
-      INTEGER IBLANK,ISTAR,IPAGE(80,200),IPAL(4),IIN(7,2),IOT,INOAI
+      CHARACTER IBLANK,ISTAR,IPAGE(80,200),IPAL(4),IIN(7,2),IOT,INOAI
      1         ,INOO,PIN(12,20)
      2         ,TITLE(80)
       COMMON /PGE/ IPAGE
@@ -901,11 +919,11 @@
      1                LPROD,IOP,IBLOW,IPCTR0)
 C     THIS THIS SUBROUTINE PRODUCES THE FUSE PLOT
       IMPLICIT INTEGER (A-Z)
-      INTEGER ISAVE(64,32),IAND,IOR,ISLASH,IBUF(8,20),IOUT(64),IOP
+      CHARACTER ISAVE(64,32),IAND,IOR,ISLASH,IBUF(8,20),IOUT(64),IOP
      1        ,IDASH,X,IBLANK,P,B
      2        ,HIFANT
 C
-     3        ,TITLE(80)
+     3        ,TITLE(*)
       LOGICAL LBUF(20),LFUSES(32,64),LDUMP,LPROD(80)
       COMMON /LUNIT/ PMS,POF,PDF
       DATA ISAVE/2048*' '/,IAND/'*'/,IOR/'+'/,ISLASH/'/'/,
@@ -930,7 +948,7 @@
       ISAVE(IPROD,32)=IAND
       RETURN
 C     PRINT FUSE PLOT
-   60 WRITE(POF,62) TITLE
+   60 WRITE(POF,62) TITLE(80)
    62 FORMAT(/,' ',80A1,//,
      1 '                11 1111 1111 2222 2222 2233',/,
      2 '    0123 4567 8901 2345 6789 0123 4567 8901',/)
@@ -980,13 +998,18 @@
 C     THIS SUBROUTINE GENERATES HEX PROGRAMMING FORMATS
       IMPLICIT INTEGER (A-Z)
       LOGICAL LFUSES(32,64)
-      INTEGER SOH,STX,ETX,BEL
-      INTEGER ZTABLE(16),ZCSUM(4),IOP,ITEMP(64)
+      CHARACTER SOH,STX,ETX,BEL
+      CHARACTER ZTABLE(16),ZCSUM(4),IOP,ITEMP(64)
+      CHARACTER H,S,IBLANK
       COMMON /LUNIT/ PMS,POF,PDF
       DATA H/'H'/,S/'S'/,IBLANK/' '/,
      1    ZTABLE/'0','1','2','3','4','5','6','7',
      2           '8','9','A','B','C','D','E','F'/
-      DATA SOH/001/,STX/002/,ETX/003/,BEL/007/
+      CHARACTER, PARAMETER :: SOHP = CHAR(1)
+      CHARACTER, PARAMETER :: STXP = CHAR(2)
+      CHARACTER, PARAMETER :: ETXP = CHAR(3)
+      CHARACTER, PARAMETER :: BELP = CHAR(7)
+      DATA SOH/SOHP/,STX/STXP/,ETX/ETXP/,BEL/BELP/
       CSUM=0
       IF(IOP.EQ.H) WRITE(PDF,10)
    10 FORMAT(//,'                                         .',//)
@@ -1038,7 +1061,7 @@
 C      FOR HIGH AND LOW PHANTOM FUSES
       IMPLICIT INTEGER (A-Z)
       LOGICAL LFUSES(32,64),LPHANT(32,64)
-      INTEGER L,C,IOT
+      CHARACTER L,C,IOT
       DATA L/'L'/,C/'C'/
       IF(ITYPE.GE.4) GO TO 20
       DO 10 IPROD=1,64
@@ -1128,7 +1151,7 @@
 C     THIS SUBROUTINE GENERATES BINARY PROGRAMMING FORMATS
 C
       IMPLICIT INTEGER (A-Z)
-      INTEGER ITEMP(4,8),H,L
+      CHARACTER ITEMP(4,8),H,L
       LOGICAL LFUSES(32,64)
       COMMON /LUNIT/ PMS,POF,PDF
       WRITE(PDF,10)
@@ -1163,7 +1186,7 @@
 C      CORRESPONDING OUTPUT PIN
       IMPLICIT INTEGER (A-Z)
       LOGICAL LFUSES(32,64)
-      INTEGER R,I1,I2,I4,I6,I8,IOT,INOAI,INOO
+      CHARACTER R,I1,I2,I4,I6,I8,IOT,INOAI,INOO
       DATA R/'R'/,I1/'1'/,I2/'2'/,I4/'4'/,I6/'6'/,I8/'8'/
       IF( (INOAI.NE.I6) .OR. (INOO.EQ.I1) .OR.  (INOO.EQ.I2) .OR.
      1    ( (IOT.EQ.R).AND.(INOO.EQ.I8) ) .OR.
@@ -1183,7 +1206,7 @@
 C      FOR HIGH AND LOW PHANTOM FUSES
       IMPLICIT INTEGER (A-Z)
 C
-      INTEGER X,IDASH,LOFANT,HIFANT, IOUT(64)
+      CHARACTER X,IDASH,LOFANT,HIFANT, IOUT(64)
       DATA X/'X'/,IDASH/'-'/,LOFANT/'0'/,HIFANT/'O'/
       DO 10 I=1,32
           IF( IOUT(I).EQ.IDASH ) IOUT(I)=HIFANT
@@ -1248,12 +1271,13 @@
      1        LFIX,LSAME,XORFND,LERR,LPHASE(20),LPHAS1(20),LBUF(20),
      2        LOUT(20),LOUTP(20),LCLOCK,LPTRST,LCTRST,LENABL(20),NREG,
      3        LSA11,LSA12,LSA01,LSA02,LPRINT
-      INTEGER  IPAGE(80,200),INOO,ISYM(8,20),ISYM1(8,20)
+      CHARACTER  IPAGE(80,200),INOO,ISYM(8,20),ISYM1(8,20)
      1         ,IDASH,L,H,X,C,Z,N0,ITEST,IPROD,ISUM,XORSUM
      2         ,N1,ERR,IBLANK,COMENT,I4,I6
      3         ,I8
      4         ,IVECTP(20),IBUF(8,20),TITLE(80)
-      INTEGER TSTVEC(20,50),ISTATE(20),IVECT(20),ISTATT(20)
+      CHARACTER NN
+      CHARACTER TSTVEC(20,50),ISTATE(20),IVECT(20),ISTATT(20)
       INTEGER BEL
       COMMON  LBLANK,LLEFT,LAND,LOR,LSLASH,LEQUAL,LRIGHT,LXOR,LXNOR
       COMMON /PGE/ IPAGE
@@ -1707,7 +1731,7 @@
       LOGICAL LBLANK,LLEFT,LAND,LOR,LSLASH,LEQUAL,LRIGHT,LXOR,LXNOR,
      1        LFIX,LPHAS1(20),LBUF(20),LCLOCK,NREG,TOR,TXOR,TXNOR,TAND,
      2        LPHASA,LPHASB
-      INTEGER L,H,X,Z,IPAGE(80,200),ISYM(8,20),ISYM1(8,20),IBUF(8,20)
+      CHARACTER L,H,X,Z,IPAGE(80,200),ISYM(8,20),ISYM1(8,20),IBUF(8,20)
      1         ,IVECT(20),IVECTP(20),ITEST,ITESTA,ITESTB
       COMMON  LBLANK,LLEFT,LAND,LOR,LSLASH,LEQUAL,LRIGHT,LXOR,LXNOR
       COMMON /PGE/ IPAGE
@@ -1774,10 +1798,17 @@
       SUBROUTINE PLOTF(ITYPE,IOT)
       IMPLICIT INTEGER (A-Z)
       LOGICAL  LFUSES (32,64),LPHANT(32,64)
-      INTEGER ZERO,ONE,IPBUF(32),IOT,STAR
-      INTEGER  TSTVEC(20,50)
-      INTEGER ISUM(4),IADR,STX,ETX,IDEC(4),IPT,IINP,J1,J2,J3,PINOUT
-      INTEGER IDECIO(4),ISUMIO(4),BUFIO(32),NFUSE,NTEST
+      CHARACTER ZERO,ONE,IOT,STAR
+      CHARACTER IPBUF(32)
+      CHARACTER NN,N1,N0,Z,C,X,H,L
+      CHARACTER TSTVEC(20,50)
+      INTEGER  ISUM(4),IADR,J3,PINOUT
+      INTEGER  IPT,IINP,J2,IDEC(4)
+      INTEGER J1
+      CHARACTER STX,ETX
+      CHARACTER IDECIO(4),ISUMIO(4),BUFIO(32)
+      INTEGER NFUSE,NTEST
+      CHARACTER ICONV
       COMMON  /LUNIT/ PMS,POF,PDF
       COMMON  /IPT/ IPT
       COMMON  /LFUSES/ LFUSES,LPHANT
@@ -1786,8 +1817,8 @@
       DATA    ZERO/'0'/, ONE/'1'/, STAR/'*'/,L/'L'/,H/'H'/,X/'X'/
      1        C/'C'/,Z/'Z'/,N0/'0'/,N1/'1'/,NN/'N'/
       IADR=0
-      STX=2
-      ETX=3
+      STX=CHAR(2)
+      ETX=CHAR(3)
       J3 =0
 C
 C           ADDITIONS  TO GENERATE PINOUT AND
@@ -1843,21 +1874,21 @@
                ISUM(2) = MOD(ISUM(2)+ISUM(4)/256,256)
                ISUM(4) = MOD(ISUM(4),256)
                DO 350 I = 1,20
-               J1 = (TSTVEC(I,J))
-               IF (J1.EQ.L) ISUM(4) = ISUM(4) + 76
-               IF (J1.EQ.H) ISUM(4) = ISUM(4) + 72
-               IF (J1.EQ.C) ISUM(4) = ISUM(4) + 67
-               IF (J1.EQ.X) ISUM(4) = ISUM(4) + 88
-               IF (J1.EQ.Z) ISUM(4) = ISUM(4) + 90
-               IF (J1.EQ.NN) ISUM(4) = ISUM(4) + 78
-               IF (J1.EQ.N1) ISUM(4) = ISUM(4) + 49
-               IF (J1.EQ.N0) ISUM(4) = ISUM(4) + 48
+               J1 = ICHAR(TSTVEC(I,J))
+               IF (J1.EQ.ICHAR(L)) ISUM(4) = ISUM(4) + 76
+               IF (J1.EQ.ICHAR(H)) ISUM(4) = ISUM(4) + 72
+               IF (J1.EQ.ICHAR(C)) ISUM(4) = ISUM(4) + 67
+               IF (J1.EQ.ICHAR(X)) ISUM(4) = ISUM(4) + 88
+               IF (J1.EQ.ICHAR(Z)) ISUM(4) = ISUM(4) + 90
+               IF (J1.EQ.ICHAR(NN)) ISUM(4) = ISUM(4) + 78
+               IF (J1.EQ.ICHAR(N1)) ISUM(4) = ISUM(4) + 49
+               IF (J1.EQ.ICHAR(N0)) ISUM(4) = ISUM(4) + 48
                ISUM(2) = MOD(ISUM(2)+ISUM(4)/256,256)
                ISUM(4) = MOD(ISUM(4),256)
   350          CONTINUE
                DO 360 I = 1,4
-               IDECIO(I) = MOD(J/10**(4-I),10)
-               ISUM(4) = ISUM(4) + IDECIO(I) + 48
+               IDECIO(I) = CHAR(MOD(J/10**(4-I),10))
+               ISUM(4) = ISUM(4) + ICHAR(IDECIO(I)) + ICHAR('0')
                ISUM(2) = MOD(ISUM(2)+ISUM(4)/256,256)
                ISUM(4) = MOD(ISUM(4),256)
   360          CONTINUE
@@ -1885,17 +1916,17 @@
       SUBROUTINE SUMCHK
       IMPLICIT INTEGER (A-Z)
       LOGICAL LFUSES(32,64),LPHANT(32,64)
-      INTEGER BUFIO(32),NFUSE
+      INTEGER NFUSE
       INTEGER ISUM(4),IDEC(4)
-      INTEGER IPBUF(32)
+      CHARACTER BUFIO(32),IPBUF(32)
       COMMON /IPT/ IPT
       COMMON /LFUSES/ LFUSES,LPHANT
       COMMON/SUM/ ISUM,IDEC,IPBUF,BUFIO,NFUSE
       DO 50 J=1,32
       IF(LPHANT(J,IPT)) GO TO 50
-      IF(LFUSES(J,IPT)) BUFIO(J)=49
-      IF(.NOT.LFUSES(J,IPT)) BUFIO(J)=48
-      ISUM(4)=ISUM(4)+BUFIO(J)
+      IF(LFUSES(J,IPT)) BUFIO(J)='1'
+      IF(.NOT.LFUSES(J,IPT)) BUFIO(J)='0'
+      ISUM(4)=ISUM(4)+ICHAR(BUFIO(J))
       ISUM(2)=MOD(ISUM(2)+ISUM(4)/256,256)
       ISUM(4)=MOD(ISUM(4),256)
    50 CONTINUE
@@ -1916,7 +1947,7 @@
 C     THIS SUBROUTINE GENERATES INTEL HEX PROGRAMMING FORMATS
       IMPLICIT INTEGER (A-Z)
       LOGICAL LFUSES(32,64)
-      INTEGER   ZTABLE(16),IOP,ITEMP(32)
+      CHARACTER   ZTABLE(16),IOP,ITEMP(32)
       COMMON /LUNIT/ PMS,POF,PDF
       DATA
      1    ZTABLE/'0','1','2','3','4','5','6','7',
